File: day-2-pm/sandbox-java/src/main/java/calc/Evaluator.java
package calc;

// Walk the AST and return the integer result. Throws CalcException on
// division by zero. Pure integer arithmetic — no floats.
public final class Evaluator {

    private Evaluator() {
    }

    public static long evaluate(Node node) {
        switch (node.kind) {
            case Number:
                return node.value;

            case BinaryOp: {
                final long left = evaluate(node.left);
                final long right = evaluate(node.right);
                switch (node.op) {
                    case Add:
                        return left + right;
                    case Sub:
                        return left - right;
                    case Mul:
                        return left * right;
                    case Div:
                        if (right == 0) {
                            throw new CalcException("division by zero");
                        }
                        return left / right;
                }
            }
        }
        throw new CalcException("evaluator: unknown node kind");
    }
}

File: day-2-pm/sandbox-python/calc/evaluator.py
"""Tree-walking evaluator.

Walk the AST and return the integer result. Raises ``CalcError`` on division
by zero. Pure integer arithmetic — no floats.
"""

from __future__ import annotations

from . import CalcError
from .parser import BinOp, Node, NodeKind


def _truncate_div(left: int, right: int) -> int:
    """Integer division that truncates toward zero (C++ semantics)."""
    quotient = abs(left) // abs(right)
    if (left < 0) != (right < 0):
        quotient = -quotient
    return quotient


def evaluate(node: Node) -> int:
    if node.kind == NodeKind.Number:
        return node.value

    if node.kind == NodeKind.BinaryOp:
        left = evaluate(node.left)
        right = evaluate(node.right)
        if node.op == BinOp.Add:
            return left + right
        if node.op == BinOp.Sub:
            return left - right
        if node.op == BinOp.Mul:
            return left * right
        if node.op == BinOp.Div:
            if right == 0:
                raise CalcError("division by zero")
            return _truncate_div(left, right)

    raise CalcError("evaluator: unknown node kind")
